packages/core/src/api/app/FeatureFlags.tsx

```diff
@@ -0,0 +1,181 @@
+ *
+ * This acts as a data structure for the user's feature flags. You
+ * can use this to retrieve, add, edit, delete, clear and save the user's
+ * feature flags to the local browser for persisted storage.
+ */
+export class UserFlags extends Map<FeatureFlagName, FeatureFlagState> {
+  static load(): UserFlags {
+    validateBrowserCompat();
+
+    try {
+      const jsonString = window.localStorage.getItem('featureFlags') as string;
+      const json = JSON.parse(jsonString);
+      return new this(Object.entries(json));
+    } catch (err) {
+      return new this([]);
+    }
+  }
+
+  get(name: FeatureFlagName): FeatureFlagState {
+    return super.get(name) || FeatureFlagState.Off;
+  }
+
+  set(name: FeatureFlagName, state: FeatureFlagState): this {
+    validateFlagName(name);
+    const output = super.set(name, state);
+    this.save();
+    return output;
+  }
+
+  delete(name: FeatureFlagName): boolean {
+    const output = super.delete(name);
+    this.save();
+    return output;
+  }
+
+  clear(): void {
+    super.clear();
+    this.save();
+  }
+
+  save(): void {
+    window.localStorage.setItem(
+      'featureFlags',
+      JSON.stringify(this.toObject()),
+    );
+  }
+
+  toObject() {
+    return Array.from(this.entries()).reduce(
+      (obj, [key, value]) => ({ ...obj, [key]: value }),
+      {},
+    );
+  }
+
+  toJSON() {
+    return JSON.stringify(this.toObject());
+  }
+
+  toString() {
+    return this.toJSON();
+  }
+}
+
+/**
+ * The FeatureFlagsRegistry class.
+ *
+ * This acts as a holding data structure for feature flags
+ * that plugins wish to register for use in Backstage.
+ */
+export interface FeatureFlagsRegistryItem {
+  pluginId: string;
+  name: FeatureFlagName;
+}
+
+export class FeatureFlagsRegistry extends Array<FeatureFlagsRegistryItem> {
+  static from(entries: FeatureFlagsRegistryItem[]) {
+    Array.from(entries).forEach(entry => validateFlagName(entry.name));
+    return new FeatureFlagsRegistry(...entries);
+  }
+
+  push(...entries: FeatureFlagsRegistryItem[]): number {
+    Array.from(entries).forEach(entry => validateFlagName(entry.name));
+    return super.push(...entries);
+  }
+
+  concat(
+    ...entries: (
+      | FeatureFlagsRegistryItem
+      | ConcatArray<FeatureFlagsRegistryItem>
+    )[]
+  ): FeatureFlagsRegistryItem[] {
+    const _concat = super.concat(...entries);
+    Array.from(_concat).forEach(entry => validateFlagName(entry.name));
+    return _concat;
+  }
+
+  toObject() {
+    return [...this.values()];
+  }
+
+  toJSON() {
+    return JSON.stringify(this.toObject());
+  }
+
+  toString() {
+    return this.toJSON();
+  }
+}
+
+/**
+ * Create the FeatureFlags implementation based on the API.
+ */
+export class FeatureFlags implements FeatureFlagsApi {
+  public registeredFeatureFlags: FeatureFlagsRegistryItem[] = [];
+  private userFlags: UserFlags | undefined;
+
+  getFlags(): UserFlags {
+    if (!this.userFlags) this.userFlags = UserFlags.load();
```

Comment:
FYI @spotify/backstage-core the reason I'm doing this instead of `private readonly userFlags: UserFlags = new UserFlags.load()` is an ES5 <-> ES6 limitation. It'll work if we change the `target` to `es6` (instead of `es5`) in `tsconfig.json` for `/packages/app/`. Otherwise this error exists:

```
$ backstage-cli test
 FAIL  src/App.test.tsx
  ● Test suite failed to run

    TypeError: Constructor Map requires 'new'
        at UserFlags.Map (<anonymous>)

      44 |     delete(name) {
      45 |         const output = super.delete(name);
    > 46 |         this.save();
         |                                          ^
      47 |         return output;
      48 |     }
      49 |     clear() {

      at new UserFlags (../../core/dist/cjs/api/app/FeatureFlags.js:46:42)
      at Function.Object.<anonymous>.UserFlags.load (../../core/dist/cjs/api/app/FeatureFlags.js:32:20)
      at new FeatureFlags (../../core/dist/cjs/api/app/FeatureFlags.js:95:36)
      at Object.<anonymous> (apis.ts:31:33)
      at Object.<anonymous> (App.tsx:24:1)
      at Object.<anonymous> (App.test.tsx:23:29)

  console.warn ../../node_modules/react/cjs/react.development.js:315
    Warning: React.createFactory() is deprecated and will be removed in a future major release. Consider using JSX or use React.createElement() directly instead.

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        3.568s, estimated 25s
Ran all test suites.
Command 'web-scripts' exited with code 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Reply:
Ah, ok, we should look into getting that fixed!